### DISCUSSION TITLE
export template formating with jinja

### DISCUSSION THREAD

**Discussion by nikunj-tele** [netbox-community/netbox#15105] [General]
Hi,
I am trying to create export template for our EOL devices list but netbox always give proxy error because  of long list which cannot be handle by netbox so , is it possible to limit the ouput with number of lines  in the csv file with jinja2 template?

2) Secondly, I want to get  the report only for one site then I do not get the output with following 
**{% for device in queryset %}
{% if device.sites == "eu" %}
---
{% endif %}
{% endfor %}**

and finally when I download csv it give the output with one line spacing instead so is there any way to remove the blank line in CSV (null line while)
<img width="430" alt="image" src="https://github.com/netbox-community/netbox/assets/55275995/5c089d40-ab45-46bb-a139-e082f5ecef3b">

Sorry I have asked multiple queries in one quiestion...

**Comment by candlerb** [comment]
> proxy error because of long list which cannot be handle by netbox

Probably your timeout from gunicorn or apache/nginx is too short, and these can be increased.

If you want to reduce the number of devices in your export, there are many approaches, including:
1. Select fewer objects to be exported (you can do this in the GUI using "Filter" tab, and there's no need to change your export)
2. Apply an offset/limit in the SQL query: `{% for device in queryset[0:50] %}`
3. Compare `loop.index` inside your for loop. See Jinja2 template designer's documentation [here](https://jinja.palletsprojects.com/en/3.1.x/templates/#for)

Alternatively, you could change from using export templates or reports to scripts. Scripts work in the background, and the browser polls every few seconds until they are completed. I think I saw that merging Reports and Scripts is being considered for a future version.

> Secondly, I want to get the report only for one site then I do not get the output with following
> `{% if device.sites == "eu" %}`

A device doesn't have an attribute `sites`, it has an attribute `site`. That gives an object, not a string, so the two won't match. I would guess that you need `device.site.name == "eu"`

The best way to test these attributes interactively is in [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/)

```
# /opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py nbshell
...
>>> d = Device.objects.get(name="nuc1")
>>> d.site
<Site: Test>
>>> d.site.name
'Test'
>>> d.site.name=='Test'
True
```

However in this case there's a better approach: filter the list of devices top be exported.
1. The user can do this in the GUI, by selecting a filter before exporting
2. You can filter the queryset - e.g. `{% for device in devices.filter(site__name="eu") %}`. Note double underscore. Again you can test this in nbshell.
3. You could use a Script instead of an Export Template/Report, and the Script can prompt the user for which Site to export. For an example see [this script](https://github.com/netbox-community/customizations/blob/master/scripts/power_summary.py).

> when I download csv it give the output with one line spacing

Again, see the template designer documentation linked above, especially the section on [whitespace control](https://jinja.palletsprojects.com/en/3.1.x/templates/#whitespace-control).

You probably want:

```
{%- for blah %}
blah
{%- endfor %}
```

**Reply by nikunj-tele** [reply]
Thanks for your help
Unfortunately , I dont have backend access but will get it sooner so I can access nbshell

**Reply by candlerb** [reply]
Aside: you don't need to run nbshell on a production instance of Netbox, you can do it on your own test installation (e.g. a VM)

**Reply by nikunj-tele** [reply]
Thanks, I am working on setting up netbox-docker instance as well :)
